### PR TITLE
Progress view chart

### PR DESCRIPTION
**Pull request by melnelen**
The data now comes from the MockData Service and the Unit tests have been added.

### CODE REVIEW

**Comment by melnelen** [comment]
Simulator Screenshot - iPhone 14 Pro - 2023-07-15 at 10 09 08

**Comment by melnelen** [comment]
Simulator Screenshot - iPhone 14 Pro - 2023-07-15 at 23 01 48

File: .DS_Store
Comment (cyril-lepottier) [inline .DS_Store:1]: To remove as the file is not related to your project.

It is generally advised to have such file in the .gitignore or even better to a [global .gitignore](https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/) as this is related to your system and not the project.

File: FocusOn/.DS_Store
Comment (cyril-lepottier) [inline FocusOn/.DS_Store:1]: To remove as well for the same reasons that above

**Review by cyril-lepottier — APPROVED** [review]

**Comment by melnelen** [comment]
The data now comes from the MockData Service and the Unit tests have been added.

Simulator Screenshot - iPhone 14 Pro - 2023-08-05 at 22 21 37

File: FocusOn/Services/DataServiceProtocol.swift [hunk 1]
```diff
@@ -6,6 +6,7 @@
 //
 
 import Foundation
+import SwiftUICharts
```
Comment (cyril-lepottier) [inline FocusOn/Services/DataServiceProtocol.swift]: Theoretically, your data services should not depend of a UI libraries such as `SwiftUICharts`.
This is thanks to the parsing done by the view models, it prevents the UI to know about `CoreData`, `API` or anything too technical, and it prevents the services to know about `SwiftUI`, `UIKit` or any anything too technical.

File: FocusOn/.DS_Store
Comment (cyril-lepottier) [reply FocusOn/.DS_Store:1]: Watch out, adding files to .gitignore doesn't remove the ones already committed, you have to remove them manually

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            let barHight = Double(calculateNumberOfCompletedTasks(goal: goal) + 1)
+            let dayComponent = Calendar.current.component(.day, from: goal.createdAt)
+            let label = LocalizedStringKey(String(dayComponent))
+            let legend = calculateGoalProgress(goal: goal)
+            
+            let dataPoint = DataPoint(value: barHight, label: label, legend: legend)
+            chartData.append(dataPoint)
+        }
+        return chartData.isEmpty ? nil : chartData
+    }
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift]: Good 👍

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+//
+//  Created by Alexandra Ivanova on 12/07/2023.
+//
+
+import Foundation
+import SwiftUICharts
+import SwiftUI
+
+class ProgressViewModel: ObservableObject {
+    @Published var allGoals: [Goal]?
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift:13]: The next iteration is to have this property `private` and to use `MockDataService` to input your test data to your unit test.

Your unit tests will look like:
```
func testSomething() {

    let mockDataService = MockDataService()
    mockDataService.allGoals = <your test data>

    let viewModel = ProgressViewModel(dataService: mockDataService)

    viewModel.executeMethodToTest()

    XCTAssert(<data to assert>)
}

```

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    let noGoal = Legend(color: .gray, label: "No goal", order: 1)
+    let fail = Legend(color: .red, label: "Fail", order: 2)
+    let smallProgress = Legend(color: .orange, label: "Small Progress", order: 3)
+    let bigProgress = Legend(color: .yellow, label: "Big Progress", order: 4)
+    let success = Legend(color: .green, label: "Success", order: 5)
+    
+    init( dataService: DataServiceProtocol = MockDataService()) {
+        self.dataService = dataService
+        self.allGoals = dataService.allGoals
+        self.chartGoalsData = [DataPoint]() // FIX
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift]: Probably not needed, the job seems done by `fillChartData()`

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    }
+    
+    func fillChartData() -> [DataPoint]? {
+        guard let unwrappedGoals = allGoals else {
+            return nil
+        }
+        var chartData: [DataPoint] = []
+        
+        for goal in unwrappedGoals {
+            let barHight = Double(calculateNumberOfCompletedTasks(goal: goal) + 1)
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift]: Typo: barHeight

File: FocusOnTests/ProgressViewModelTests.swift [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    
+    func testFillChartDataWithGoals() {
+        // Given
+        let mockGoals = viewModel.allGoals!
+        // When
+        let chartData = viewModel.fillChartData()
+        // Then
+        XCTAssertNotNil(chartData)
+        XCTAssertEqual(chartData?.count, mockGoals.count)
```
Comment (cyril-lepottier) [inline FocusOnTests/ProgressViewModelTests.swift]: I would suggest to add a bit more test such as "expected bar height"

**Review by cyril-lepottier — APPROVED** [review]
It's good, the unit test have a good shape

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            legendSignature = smallProgress
+        case 2:
+            legendSignature = bigProgress
+        case 3:
+            legendSignature = success
+        default:
+            legendSignature = noGoal
+        }
+        return legendSignature
+    }
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift]: Those 2 methods should be private and should be tested through `fillChartData()`

File: FocusOn/View/ProgressView.swift [hunk 8]
```diff
@@ -6,15 +6,42 @@
...
+                BarChartView(dataPoints: chartData ?? [])
+                    .chartStyle(
+                        BarChartStyle(
+                            barMinHeight: 10,
+                            showAxis: false,
+                            showLegends: true
+                        )
+                    )
+            }
+            .padding(20)
```
Comment (cyril-lepottier) [inline FocusOn/View/ProgressView.swift]: You can have implement a pagination style thanks to to the style `page` of the component `TabView`, [here](https://www.hackingwithswift.com/quick-start/swiftui/how-to-create-scrolling-pages-of-content-using-tabviewstyle) is a nice tutorial that should help.

If it doesn't, here is a piece of code that works:
```
Text("Here is your progres")
    .foregroundColor(.accentColor)
    .font(.largeTitle)

let chunkedChartData = chartData.chunked(into: 7)
let numberOfWeek = chunkedChartData.count
TabView {
    ForEach(0..<numberOfWeek, id: \.self) { weekNumber in
        ProgressPageView(
            title: "Week \(weekNumber)",
            dataPoint: chunkedChartData[weekNumber]
        )
    }
}
.tabViewStyle(.page)

private struct ProgressPageView: View {

    let title: String
    let dataPoint: [DataPoint]

    var body: some View {
        VStack(
            alignment: .leading,
            spacing: 20
        ) {
            Text(self.title)
                .foregroundColor(.accentColor)
                .font(.largeTitle)
            HStack {
                BarChartView(dataPoints: self.dataPoint)
                    .chartStyle(
                        BarChartStyle(
                            barMinHeight: 10,
                            showAxis: false,
                            showLegends: false
                        )
                    )
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
            }
            .padding(20)
        }
    }
}
```

Note: in the above code, I created a custom view in order to reduce the code size of `ProgressView.body`, this is not mandatory, but I advice you to do the same and to try to take the habit of reducing code size.

File: FocusOnTests/ProgressViewModelTests.swift (hunk 6 above)
Comment (cyril-lepottier) [reply FocusOnTests/ProgressViewModelTests.swift]: So basically trying to merge this test with the one of `calculateNumberOfCompletedTasks(goal:)` and the one of `calculateGoalProgress(goal:).

At the end, the test should look like this piece of code I did, with eventually even more assert:
```
// Given
let mockGoals: [Goal] = [
    Goal(
        name: "Test goal 2",
        createdAt: Calendar.current.date(byAdding: .day, value: -2, to: Date())!,
        isCompleted: false,
        tasks: [
            Task(name: "Test task 2.1", isCompleted: true),
            Task(name: "Test task 2.2", isCompleted: false),
            Task(name: "Test task 2.3", isCompleted: true)
        ]
    ),
    Goal(
        name: "Test goal 1",
        createdAt: Calendar.current.date(byAdding: .day, value: -1, to: Date())!,
        isCompleted: true,
        tasks: [
            Task(name: "Test task 1.1", isCompleted: true),
            Task(name: "Test task 1.2", isCompleted: true),
            Task(name: "Test task 1.3", isCompleted: true)
        ]
    )
]
dataService.allGoals = mockGoals // viewModel.allGoals = mockGoals in your case but I suggest you to delete ProgressViewModel.allGoals

// When
let chartData = viewModel.fillChartData()

// Then
XCTAssertEqual(chartData?.count, mockGoals.count)

XCTAssertEqual(chartData?[0].legend, Legend(color: .yellow, label: "Big Progress", order: 4))
XCTAssertEqual(chartData?[0].startValue, 0)
XCTAssertEqual(chartData?[0].endValue, 3)
XCTAssertEqual(chartData?[0].visible, true)
XCTAssertEqual(chartData?[0].label, "")

XCTAssertEqual(chartData?[1].legend, Legend(color: .green, label: "Success", order: 5))
XCTAssertEqual(chartData?[1].startValue, 0)
XCTAssertEqual(chartData?[1].endValue, 4)
XCTAssertEqual(chartData?[1].visible, true)
XCTAssertEqual(chartData?[1].label, "")
```

**Comment by melnelen** [comment]
Simulator Screenshot - iPhone 14 Pro - 2023-10-12 at 22 11 28

https://github.com/melnelen/FocusOn/assets/8265393/9bd76714-6546-46b4-9325-814b9836b3ef

Polished the UI with some paddings
Make sure all weeks are made of 7 days and ensure that the weeks start on a Monday

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+            var dataPointsForWeek: [DataPoint] = []
+            
+            guard let earliestDate = weeklyGoals.map({ $0.createdAt }).min() else {
+                continue
+            }
+            
+            // Find the first Monday in the week
+            let calendar = Calendar.current
+            var components = calendar.dateComponents([.yearForWeekOfYear, .weekOfYear, .weekday], from: earliestDate)
+            components.weekday = firstWeekday // Monday
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift]: Side note: there is a development philosophy that says we shouldn't comment code because the code should be understandable by itself and bad comments are worse that no comment at all. Don't apply it, I think OC evaluators want comments, but in this case, I personally think that it would have been better to name your constant `monday` rather than having to document it via comments.

In this case, I think the better option would be to use the user preferences to let them decide what should be the first day of the week (some like Sunday), but let's leave that for another time.

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        var dataPointChunks: [[DataPoint]] = []
+        
+        for weeklyGoals in goalChunks {
+            var dataPointsForWeek: [DataPoint] = []
+            
+            guard let earliestDate = weeklyGoals.map({ $0.createdAt }).min() else {
+                continue
+            }
+            
+            // Find the first Monday in the week
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift]: I think this part is not necessary, you just want to add missing data regardless of the position

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+                goalChunks.append([goal])
+            } else {
+                goalChunks[goalChunks.count - 1].append(goal)
+            }
+        }
+        return goalChunks
+    }
+    
+    private func generateDataPointsForWeeklyChunksOf(goals: [Goal]) -> [[DataPoint]] {
+        let goalChunks = splitGoalsIntoWeeklyChunks(goals: goals)
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift:90]: I think this should be the last part of you algorithm, you want to clean your data before splitting them or at the same time eventually. But before, it will create some issue because you will have sometime to add data, meaning you will have to move one to the next split, which will one to the next split, etc. Seems like impossible to handle

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        return DataPoint(value: barHeight, label: label, legend: legend)
+    }
+    
+    private func splitGoalsIntoWeeklyChunks(goals: [Goal]) -> [[Goal]] {
+        let sortedGoals = goals.sorted { $0.createdAt < $1.createdAt }
+        var goalChunks: [[Goal]] = []
+        
+        for goal in sortedGoals {
+            let weekday = calendar.component(.weekday, from: goal.createdAt)
+            if weekday == firstWeekday || goalChunks.isEmpty {
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift:80]: What if the `firstWeekday` is missing?

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    private func generateDataPointForGoal(goal: Goal) -> DataPoint {
+        let barHeight = Double(calculateNumberOfCompletedTasks(goal: goal) + 1)
+        let dayComponent = Calendar.current.component(.day, from: goal.createdAt)
+        let label = LocalizedStringKey(String(dayComponent))
+        let legend = calculateGoalProgress(goal: goal)
+        
+        return DataPoint(value: barHeight, label: label, legend: legend)
+    }
+    
+    private func splitGoalsIntoWeeklyChunks(goals: [Goal]) -> [[Goal]] {
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift:74]: I think this method would be much simpler if you used the calendar component `weekOfYear` and `year`.

The idea  of the method is to quite simply regroup the data that belongs to the same week together.

File: FocusOn/ViewModel/ProgressViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+            components.weekday = firstWeekday // Monday
+            
+            guard let firstMonday = calendar.date(from: components) else {
+                continue
+            }
+            
+            for i in 0..<7 {
+                let currentDate = calendar.date(byAdding: .day, value: i, to: firstMonday) ?? Date()
+                
+                if let goal = weeklyGoals.first(where: { calendar.isDate($0.createdAt, inSameDayAs: currentDate) }) {
```
Comment (cyril-lepottier) [inline FocusOn/ViewModel/ProgressViewModel.swift]: Not sure what `currentDate` is for.

Here,  I think you should just check if the data exist for Monday, Tuesday, ..., Sunday, if it does, you add it to `dataPointsForWeek ` if it doesn't you create an empty data point. The component `weekday ` should do that for you.

**Review by cyril-lepottier — COMMENTED** [review]
I added some comments in `generateDataPointsForWeeklyChunksOf` and `splitGoalsIntoWeeklyChunks `. I hope they make sense and are useful.

We can also try it in a mentoring session, this kind of ViewModel code is also very interesting from the unit tests perspective, we could even have a look at TDD.
